webauthn-rs-core/src/attestation.rs

```diff
@@ -688,7 +859,337 @@ pub(crate) fn verify_apple_anonymous_attestation(
+                    }
+                }
+                openssl::pkey::Id::EC => {
+                    let ec_key = public_key.ec_key()?;
+                    let mut ctxt = openssl::bn::BigNumContext::new()?;
+                    let raw = ec_key.public_key().to_bytes(
+                        ec_key.group(),
+                        openssl::ec::PointConversionForm::UNCOMPRESSED,
+                        &mut ctxt,
+                    )?;
+
+                    match meta.algorithm() {
+                        "ES256" => ES256PublicKey::from_bytes(&raw)?.verify_token(token, opts)?,
+                        _ => return Err(SafetyNetError::BadAlg),
+                    }
+                }
+                _ => return Err(SafetyNetError::BadAlg),
+            };
+
+        // 3. Verify that the nonce attribute in the payload of response is identical to the Base64 encoding of the SHA-256 hash of the concatenation of authenticatorData and clientDataHash.
+        let nonce = verified_claims.nonce.ok_or(SafetyNetError::MissingNonce)?;
+        let nonce = base64::decode(&nonce).map_err(|_| SafetyNetError::BadNonce)?;
+        if nonce != data_to_verify.to_vec() {
+            return Err(SafetyNetError::NonceMismatch);
+        }
+
+        // 4. Verify that the SafetyNet response actually came from the SafetyNet service by following the steps in the SafetyNet online documentation.
+        let common_name = {
+            let name = cert
+                .subject_name()
+                .entries_by_nid(openssl::nid::Nid::COMMONNAME)
+                .next()
+                .ok_or(SafetyNetError::InvalidHostname)?;
+            name.data().as_utf8()?.to_string()
+        };
+
+        // §8.5.5 Verify that attestationCert is issued to the hostname "attest.android.com"
+        if common_name.as_str() != "attest.android.com" {
+            return Err(SafetyNetError::InvalidHostname);
+        }
+
+        // §8.5.6 Verify that the ctsProfileMatch attribute in the payload of response is true.
+        if !verified_claims.custom.cts_profile_match {
+            return Err(SafetyNetError::CtsProfileMatchFailed);
+        }
+
+        // Verify sanity of timestamp in the payload
+        if !danger_ignore_timestamp {
+            let expires: std::time::Duration = std::time::SystemTime::now()
+                .duration_since(std::time::UNIX_EPOCH)?
+                + std::time::Duration::from_secs(60);
+            if verified_claims.custom.timestamp_ms as u128 > expires.as_millis() {
+                return Err(SafetyNetError::Expired);
+            }
+        }
+
+        Ok((certs, verified_claims.custom))
+    }(response_str)
+    .map_err(|e| {
+        error!("jwt saftey-net error: {:?}", e);
+        WebauthnError::AttestationStatementResponseInvalid
+    })?;
+
+    // 5. If successful, return implementation-specific values representing attestation type Anonymous CA and attestation trust path x5c.
+    Ok(ParsedAttestationData::Basic(x5c))
```

Comment:
This format seems to have a lot of extra metadata, is any of it worth capturing for auditing?

Reply:
Yeah absolutely, I think I want to make ParsedAttestationData a struct to include the format, and an enum of metadata that was extracted. I want to adjust the VerifiedAttestation thing to pull out the data and do the comparison one level up so we have access to the metadata

Reply:
Okay, I think that's a good idea, but maybe a subsequent PR ?